src/libs/Middleware/Pagination.ts

```diff
@@ -0,0 +1,127 @@
+// TODO: Is this a legit use case for exposing `OnyxCache`, or should we use `Onyx.connect`?
+import fastMerge from 'expensify-common/dist/fastMerge';
+import type {OnyxCollection} from 'react-native-onyx';
+import Onyx from 'react-native-onyx';
+import type {ApiCommand} from '@libs/API/types';
+import Log from '@libs/Log';
+import PaginationUtils from '@libs/PaginationUtils';
+import CONST from '@src/CONST';
+import type {OnyxCollectionKey, OnyxPagesKey, OnyxValues} from '@src/ONYXKEYS';
+import type {Request} from '@src/types/onyx';
+import type {PaginatedRequest} from '@src/types/onyx/Request';
+import type Middleware from './types';
+
+type PaginationConfig<TResourceKey extends OnyxCollectionKey, TPageKey extends OnyxPagesKey> = {
+    initialCommand: ApiCommand;
+    previousCommand: ApiCommand;
+    nextCommand: ApiCommand;
+    resourceCollectionKey: TResourceKey;
+    pageCollectionKey: TPageKey;
+    sortItems: (items: OnyxValues[TResourceKey]) => Array<OnyxValues[TResourceKey] extends Record<string, infer TResource> ? TResource : never>;
+    getItemID: (item: OnyxValues[TResourceKey] extends Record<string, infer TResource> ? TResource : never) => string;
+};
+
+type PaginationConfigMapValue = Omit<PaginationConfig<OnyxCollectionKey, OnyxPagesKey>, 'initialCommand' | 'previousCommand' | 'nextCommand'> & {
+    type: 'initial' | 'next' | 'previous';
+};
+
+const paginationConfigs = new Map<string, PaginationConfigMapValue>();
+const ressources = new Map<OnyxCollectionKey, OnyxCollection<OnyxValues[OnyxCollectionKey]>>();
+const pages = new Map<OnyxPagesKey, OnyxCollection<OnyxValues[OnyxPagesKey]>>();
+
+function registerPaginationConfig<TResourceKey extends OnyxCollectionKey, TPageKey extends OnyxPagesKey>({
+    initialCommand,
+    previousCommand,
+    nextCommand,
+    ...config
+}: PaginationConfig<TResourceKey, TPageKey>): void {
+    // TODO: Is there a way to avoid these casts?
+    paginationConfigs.set(initialCommand, {...config, type: 'initial'} as unknown as PaginationConfigMapValue);
+    paginationConfigs.set(previousCommand, {...config, type: 'previous'} as unknown as PaginationConfigMapValue);
+    paginationConfigs.set(nextCommand, {...config, type: 'next'} as unknown as PaginationConfigMapValue);
```

Comment:
I spent a lot of time on these today. TLDR: it's won't be easy to avoid these assertions (and not sure if it's worth it)

The real issue is that the `config` from `registerPaginationConfig` params is very specific while `paginationConfigs` is on the other spectrum quite loose:
```ts
const paginationConfigs = new Map<string, PaginationConfigMapValue>();
```
☝️ Any key in this map has a very generic value – any collection value (report, reportActions, policy etc)

This results in a mismatch between `sortItems` and `getItemID` types:
![image](https://github.com/Expensify/App/assets/46095609/737cef93-0322-46e8-ad49-256ee00f4bc1)

Reply:
A possible fix that would require quite a lot changes (+ additional types) is to use a [mapped object](https://www.typescriptlang.org/docs/handbook/2/mapped-types.html) instead of `new Map()`.

Here is the approach that works:
```ts
type ApiCommandToOnyxKeyMap = {
    [WRITE_COMMANDS.OPEN_REPORT]: typeof ONYXKEYS.COLLECTION.REPORT;
    [READ_COMMANDS.GET_OLDER_ACTIONS]: typeof ONYXKEYS.COLLECTION.REPORT;
    [READ_COMMANDS.GET_NEWER_ACTIONS]: typeof ONYXKEYS.COLLECTION.REPORT;
    // ... other commands ...
};
type PaginatedCommands = keyof ApiCommandToOnyxKeyMap;
type PaginatedOnyxKeys = ValueOf<ApiCommandToOnyxKeyMap>;

type PaginationConfigMap = {
    [Command in PaginatedCommands]?: PaginationConfigMapValue<ApiCommandToOnyxKeyMap[Command]>;
};

type PaginationCommonConfig<TResourceKey extends PaginatedOnyxKeys = PaginatedOnyxKeys, TPageKey extends OnyxPagesKey = OnyxPagesKey> = {
    resourceCollectionKey: TResourceKey;
    pageCollectionKey: TPageKey;
    sortItems: (items: Array<OnyxValues[TResourceKey]>) => Array<OnyxValues[TResourceKey]>;
    getItemID: (item: OnyxValues[TResourceKey]) => string;
};

type PaginationConfig<TResourceKey extends PaginatedOnyxKeys, TPageKey extends OnyxPagesKey> = PaginationCommonConfig<TResourceKey, TPageKey> & {
    initialCommand: PaginatedCommands;
    previousCommand: PaginatedCommands;
    nextCommand: PaginatedCommands;
};

type PaginationConfigMapValue<TResourceKey extends PaginatedOnyxKeys> = PaginationCommonConfig<TResourceKey, OnyxPagesKey> & {
    type: 'initial' | 'next' | 'previous';
};

const paginationConfigs: PaginationConfigMap = {};
const ressources = new Map<OnyxCollectionKey, OnyxCollection<OnyxValues[OnyxCollectionKey]>>();
const pages = new Map<OnyxPagesKey, OnyxCollection<OnyxValues[OnyxPagesKey]>>();

function registerPaginationConfig<TResourceKey extends PaginatedOnyxKeys, TPageKey extends OnyxPagesKey>({
    initialCommand,
    previousCommand,
    nextCommand,
    sortItems,
    getItemID,
    ...config
}: PaginationConfig<TResourceKey, TPageKey>): void {
    paginationConfigs[initialCommand] = {...config, sortItems, getItemID, type: 'initial'};
```

Reply:
cool, let's :donothing: for now